File: iOS/TalkingDataHTML.m
//
//  TalkingDataHTML.m
//  TalkingData-HTML
//
//  Created by liweiqiang on 14-1-9.
//  Copyright (c) 2014年 tendcloud. All rights reserved.
//

#import "TalkingDataHTML.h"
#import <UIKit/UIKit.h>
#import <WebKit/WebKit.h>
#import "TalkingData.h"


@interface TalkingDataHTML ()

@property (nonatomic, strong) NSString *currPageName;

@end


@implementation TalkingDataHTML

+ (BOOL)execute:(NSURL *)url webView:(WKWebView *)webView {
    NSString *parameters = [url.absoluteString stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    if ([parameters hasPrefix:@"talkingdata"]) {
        static TalkingDataHTML *talkingDataHTML = nil;
        static dispatch_once_t predicate;
        dispatch_once(&predicate, ^{
            talkingDataHTML = [[TalkingDataHTML alloc] init];
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
            if ([TalkingData respondsToSelector:@selector(setFrameworkTag:)]) {
                [TalkingData performSelector:@selector(setFrameworkTag:) withObject:@6];
            }
#pragma clang diagnostic pop
        });
        NSString *str = [parameters substringFromIndex:12];
        NSDictionary *dic = [talkingDataHTML jsonToDictionary:str];
        NSString *functionName = [dic objectForKey:@"functionName"];
        NSArray *args = [dic objectForKey:@"arguments"];
        if ([functionName isEqualToString:@"getDeviceId"]) {
            [talkingDataHTML getDeviceId:args webView:webView];
        } else {
            SEL selector = NSSelectorFromString([NSString stringWithFormat:@"%@:", functionName]);
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            if ([talkingDataHTML respondsToSelector:selector]) {
                [talkingDataHTML performSelector:selector withObject:args];
            }
#pragma clang diagnostic pop
        }
        return YES;
    }
    
    return NO;
}

- (void)getDeviceId:(NSArray *)arguments webView:(WKWebView *)webView {
    NSString *arg0 = [arguments objectAtIndex:0];
    if (arg0 == nil || [arg0 isKindOfClass:[NSNull class]] || arg0.length == 0) {
        return;
    }
    NSString *deviceId = [TalkingData getDeviceID];
    NSString *callBack = [NSString stringWithFormat:@"%@('%@')", arg0, deviceId];
    if ([webView isKindOfClass:[WKWebView class]]) {
        [webView evaluateJavaScript:callBack completionHandler:nil];
    }
}

- (void)setLocation:(NSArray *)arguments {
    NSString *arg0 = [arguments objectAtIndex:0];
    if (arg0 == nil || [arg0 isKindOfClass:[NSNull class]]) {
        return;
    }
    NSString *arg1 = [arguments objectAtIndex:1];
    if (arg1 == nil || [arg1 isKindOfClass:[NSNull class]]) {
        return;
    }
    double latitude = [arg0 doubleValue];
    double longitude = [arg1 doubleValue];
    [TalkingData setLatitude:latitude longitude:longitude];
}

- (void)onRegister:(NSArray *)arguments {
    NSString *profileId = [arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    NSUInteger type = [[arguments objectAtIndex:1] unsignedIntegerValue];
    NSString *name = [arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    [TalkingData onRegister:profileId type:type name:name];
}

- (void)onLogin:(NSArray *)arguments {
    NSString *profileId = [arguments objectAtIndex:0];
    if (![profileId isKindOfClass:[NSString class]]) {
        profileId = nil;
    }
    NSUInteger type = [[arguments objectAtIndex:1] unsignedIntegerValue];
    NSString *name = [arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    [TalkingData onLogin:profileId type:type name:name];
}

- (void)onEvent:(NSArray *)arguments {
    NSString *eventId = [arguments objectAtIndex:0];
    if (eventId == nil || [eventId isKindOfClass:[NSNull class]]) {
        return;
    }
    [TalkingData trackEvent:eventId];
}

- (void)onEventWithLabel:(NSArray *)arguments {
    NSString *eventId = [arguments objectAtIndex:0];
    if (eventId == nil || [eventId isKindOfClass:[NSNull class]]) {
        return;
    }
    NSString *eventLabel = [arguments objectAtIndex:1];
    if ([eventLabel isKindOfClass:[NSNull class]]) {
        eventLabel = nil;
    }
    [TalkingData trackEvent:eventId label:eventLabel];
}

- (void)onEventWithParameters:(NSArray *)arguments {
    NSString *eventId = [arguments objectAtIndex:0];
    if (eventId == nil || [eventId isKindOfClass:[NSNull class]]) {
        return;
    }
    NSString *eventLabel = [arguments objectAtIndex:1];
    if (eventLabel == nil || [eventLabel isKindOfClass:[NSNull class]]) {
        eventLabel = nil;
    }
    NSDictionary *parameters = [arguments objectAtIndex:2];
    if (parameters == nil && [parameters isKindOfClass:[NSNull class]]) {
        parameters = nil;
    }
    [TalkingData trackEvent:eventId label:eventLabel parameters:parameters];
}

- (void)onEventWithValue:(NSArray *)arguments {
    NSString *eventId = [arguments objectAtIndex:0];
    if (eventId == nil || [eventId isKindOfClass:[NSNull class]]) {
        return;
    }
    NSString *eventLabel = [arguments objectAtIndex:1];
    if (eventLabel == nil || [eventLabel isKindOfClass:[NSNull class]]) {
        eventLabel = nil;
    }
    NSDictionary *parameters = [arguments objectAtIndex:2];
    if (parameters == nil || [parameters isKindOfClass:[NSNull class]]) {
        parameters = nil;
    }
    NSNumber *eventValue = [arguments objectAtIndex:3];
    if (eventValue == nil || [eventValue isKindOfClass:[NSNull class]]) {
        eventValue = nil;
    }
    [TalkingData trackEvent:eventId label:eventLabel parameters:parameters value:[eventValue doubleValue]];
}

- (void)onPlaceOrder:(NSArray *)arguments {
    NSString *orderId = [arguments objectAtIndex:0];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    NSNumber *amount = [arguments objectAtIndex:1];
    if (amount == nil || ![amount isKindOfClass:[NSNumber class]]) {
        amount = nil;
    }
    NSString *currencyType = [arguments objectAtIndex:2];
    if (![currencyType isKindOfClass:[NSString class]]) {
        currencyType = nil;
    }
    [TalkingData onPlaceOrder:orderId amount:[amount intValue] currencyType:currencyType];
}

- (void)onOrderPaySucc:(NSArray *)arguments {
    NSString *orderId = [arguments objectAtIndex:0];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    NSNumber *amount = [arguments objectAtIndex:1];
    if (amount == nil || ![amount isKindOfClass:[NSNumber class]]) {
        amount = nil;
    }
    NSString *currencyType = [arguments objectAtIndex:2];
    if (![currencyType isKindOfClass:[NSString class]]) {
        currencyType = nil;
    }
    NSString *paymentType = [arguments objectAtIndex:3];
    if (![paymentType isKindOfClass:[NSString class]]) {
        paymentType = nil;
    }
    [TalkingData onOrderPaySucc:orderId amount:[amount intValue] currencyType:currencyType paymentType:paymentType];
}

- (void)onCancelOrder:(NSArray *)arguments {
    NSString *orderId = [arguments objectAtIndex:0];
    if (![orderId isKindOfClass:[NSString class]]) {
        orderId = nil;
    }
    NSNumber *amount = [arguments objectAtIndex:1];
    if (amount == nil || ![amount isKindOfClass:[NSNumber class]]) {
        amount = nil;
    }
    NSString *currencyType = [arguments objectAtIndex:2];
    if (![currencyType isKindOfClass:[NSString class]]) {
        currencyType = nil;
    }
    [TalkingData onCancelOrder:orderId amount:[amount intValue] currencyType:currencyType];
}

- (void)onViewItem:(NSArray *)arguments {
    NSString *itemId = [arguments objectAtIndex:0];
    if (![itemId isKindOfClass:[NSString class]]) {
        itemId = nil;
    }
    NSString *category = [arguments objectAtIndex:1];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    NSString *name = [arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    int unitPrice = [[arguments objectAtIndex:3] intValue];
    [TalkingData onViewItem:itemId category:category name:name unitPrice:unitPrice];
}

- (void)onAddItemToShoppingCart:(NSArray *)arguments {
    NSString *itemId = [arguments objectAtIndex:0];
    if (![itemId isKindOfClass:[NSString class]]) {
        itemId = nil;
    }
    NSString *category = [arguments objectAtIndex:1];
    if (![category isKindOfClass:[NSString class]]) {
        category = nil;
    }
    NSString *name = [arguments objectAtIndex:2];
    if (![name isKindOfClass:[NSString class]]) {
        name = nil;
    }
    int unitPrice = [[arguments objectAtIndex:3] intValue];
    int amount = [[arguments objectAtIndex:4] intValue];
    [TalkingData onAddItemToShoppingCart:itemId category:category name:name unitPrice:unitPrice amount:amount];
}

- (void)onViewShoppingCart:(NSArray *)arguments {
    NSDictionary *shoppingCartDic = [arguments objectAtIndex:0];
    if (![shoppingCartDic isKindOfClass:[NSDictionary class]]) {
        return;
    }
    TalkingDataShoppingCart *shoppingCart = [self shoppingCartFromDictionary:shoppingCartDic];
    [TalkingData onViewShoppingCart:shoppingCart];
}

- (void)onPage:(NSArray *)arguments {
    NSString *pageName = [arguments objectAtIndex:0];
    if (pageName == nil || [pageName isKindOfClass:[NSNull class]]) {
        return;
    }
    
    if (self.currPageName) {
        [TalkingData trackPageEnd:self.currPageName];
    }
    self.currPageName = pageName;
    [TalkingData trackPageBegin:self.currPageName];
}

- (void)onPageBegin:(NSArray *)arguments {
    NSString *pageName = [arguments objectAtIndex:0];
    if (pageName == nil || [pageName isKindOfClass:[NSNull class]]) {
        return;
    }
    self.currPageName = pageName;
    [TalkingData trackPageBegin:pageName];
}

- (void)onPageEnd:(NSArray *)arguments {
    NSString *pageName = [arguments objectAtIndex:0];
    if (pageName == nil || [pageName isKindOfClass:[NSNull class]]) {
        return;
    }
    [TalkingData trackPageEnd:pageName];
    self.currPageName = nil;
}

- (NSDictionary *)jsonToDictionary:(NSString *)jsonStr {
    if (jsonStr) {
        NSError *error = nil;
        id object = [NSJSONSerialization JSONObjectWithData:[jsonStr dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingAllowFragments error:&error];
        if (error == nil && [object isKindOfClass:[NSDictionary class]]) {
            return object;
        }
    }
    
    return nil;
}

- (TalkingDataOrder *)orderFormDictionary:(NSDictionary *)orderDic {
    TalkingDataOrder *order = [TalkingDataOrder createOrder:orderDic[@"orderId"] total:[orderDic[@"total"] intValue] currencyType:orderDic[@"currencyType"]];
    NSArray *items = orderDic[@"items"];
    for (NSDictionary *item in items) {
        [order addItem:item[@"itemId"] category:item[@"category"] name:item[@"name"] unitPrice:[item[@"unitPrice"] intValue] amount:[item[@"amount"] intValue]];
    }
    
    return order;
}

- (TalkingDataShoppingCart *)shoppingCartFromDictionary:(NSDictionary *)shoppingCartDic {
    TalkingDataShoppingCart *shoppingCart = [TalkingDataShoppingCart createShoppingCart];
    NSArray *items = shoppingCartDic[@"items"];
    for (NSDictionary *item in items) {
        [shoppingCart addItem:item[@"itemId"] category:item[@"category"] name:item[@"name"] unitPrice:[item[@"unitPrice"] intValue] amount:[item[@"amount"] intValue]];
    }
    
    return shoppingCart;
}

@end

File: Demo/Android/app/src/main/java/com/talkingdata/demo/TalkingDataHTML.java
package com.talkingdata.demo;

import android.app.Activity;
import android.webkit.WebView;

import com.tendcloud.tenddata.Order;
import com.tendcloud.tenddata.ShoppingCart;
import com.tendcloud.tenddata.TCAgent;
import com.tendcloud.tenddata.TDProfile;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class TalkingDataHTML {
    private static volatile TalkingDataHTML talkingDataHTML = null;
    
    public static TalkingDataHTML GetInstance() {
        if (talkingDataHTML == null) {
            synchronized (TalkingDataHTML.class) {
                if (talkingDataHTML == null) {
                    talkingDataHTML = new TalkingDataHTML();
                }
            }
        }
        return talkingDataHTML;
    }
    
    Activity activity = null;
    String currPageName = null;
    
    public void execute(final Activity activity, final String url, final WebView webView) throws Exception {
        if (url.startsWith("talkingdata")) {
            talkingDataHTML.activity = activity;
            String str = url.substring(12);
            JSONObject jsonObj = new JSONObject(str);
            String functionName = jsonObj.getString("functionName");
            JSONArray args = jsonObj.getJSONArray("arguments");
            if (functionName.equals("getDeviceId")) {
                talkingDataHTML.getDeviceId(args, webView);
            }else if(functionName.equals("getOAID")){
                talkingDataHTML.getOAID(args, webView);
            } else {
                Class<TalkingDataHTML> classType = TalkingDataHTML.class;
                Method method = classType.getDeclaredMethod(functionName, JSONArray.class);
                method.invoke(talkingDataHTML, args);
            }
        }
    }
    
    private void getDeviceId(final JSONArray args, final WebView webView) throws JSONException {
        String deviceId = TCAgent.getDeviceId(MainActivity.context);
        String callBack = args.getString(0);
        webView.loadUrl("javascript:" + callBack + "('" + deviceId + "')");
    }
    private void getOAID(final JSONArray args, final WebView webView) throws JSONException {
        String oaid = TCAgent.getOAID(MainActivity.context);
        String callBack = args.getString(0);
        webView.loadUrl("javascript:" + callBack + "('" + oaid + "')");
    }
    
    @SuppressWarnings("unused")
    private void onRegister(final JSONArray args) throws JSONException {
        String profileId = args.getString(0);
        TDProfile.ProfileType type = this.intToProfileType(args.getInt(1));
        String name = args.getString(2);
        TCAgent.onRegister(profileId, type, name);
    }
    
    @SuppressWarnings("unused")
    private void onLogin(final JSONArray args) throws JSONException {
        String profileId = args.getString(0);
        TDProfile.ProfileType type = this.intToProfileType(args.getInt(1));
        String name = args.getString(2);
        TCAgent.onLogin(profileId, type, name);
    }
    
    @SuppressWarnings("unused")
    private void onEvent(final JSONArray args) throws JSONException {
        String eventId = args.getString(0);
        TCAgent.onEvent(MainActivity.context, eventId);
    }
    
    @SuppressWarnings("unused")
    private void onEventWithLabel(final JSONArray args) throws JSONException {
        String eventId = args.getString(0);
        String eventLabel = args.getString(1);
        TCAgent.onEvent(MainActivity.context, eventId, eventLabel);
    }
    
    @SuppressWarnings("unused")
    private void onEventWithParameters(final JSONArray args) throws JSONException {
        String eventId = args.getString(0);
        String eventLabel = args.getString(1);
        String eventDataJson = args.getString(2);
        Map<String, Object> eventData = this.toMap(eventDataJson);
        TCAgent.onEvent(MainActivity.context, eventId, eventLabel, eventData);
    }
    
    @SuppressWarnings("unused")
    private void onEventWithValue(final JSONArray args) throws JSONException {
        String eventId = args.getString(0);
        String eventLabel = args.getString(1);
        String eventDataJson = args.getString(2);
        Map<String, Object> eventData = this.toMap(eventDataJson);
        double eventValue = args.getDouble(3);
        TCAgent.onEvent(MainActivity.context, eventId, eventLabel, eventData, eventValue);
    }
    
    @SuppressWarnings("unused")
    private void onPlaceOrder(final JSONArray args) throws JSONException {
        String profileId = args.getString(0);
        JSONObject orderJson = args.getJSONObject(1);
        Order order = this.orderFromJSONObject(orderJson);
        TCAgent.onPlaceOrder(profileId, order);
    }
    
    @SuppressWarnings("unused")
    private void onOrderPaySucc(final JSONArray args) throws JSONException {
        String profileId = args.getString(0);
        String payType = args.getString(1);
        JSONObject orderJson = args.getJSONObject(2);
        Order order = this.orderFromJSONObject(orderJson);
        TCAgent.onOrderPaySucc(profileId, payType, order);
    }
    
    @SuppressWarnings("unused")
    private void onViewItem(final JSONArray args) throws JSONException {
        String itemId = args.getString(0);
        String category = args.getString(1);
        String name = args.getString(2);
        int unitPrice = args.getInt(3);
        TCAgent.onViewItem(itemId, category, name, unitPrice);
    }
    
    @SuppressWarnings("unused")
    private void onAddItemToShoppingCart(final JSONArray args) throws JSONException {
        String itemId = args.getString(0);
        String category = args.getString(1);
        String name = args.getString(2);
        int unitPrice = args.getInt(3);
        int amount = args.getInt(4);
        TCAgent.onAddItemToShoppingCart(itemId, category, name, unitPrice, amount);
    }
    
    @SuppressWarnings("unused")
    private void onViewShoppingCart(final JSONArray args) throws JSONException {
        JSONObject shoppingCartJson = args.getJSONObject(0);
        ShoppingCart shoppingCart = this.shoppingCartFromJSONObject(shoppingCartJson);
        TCAgent.onViewShoppingCart(shoppingCart);
    }
    
    @SuppressWarnings("unused")
    private void onPage(final JSONArray args) throws JSONException {
        String pageName = args.getString(0);
        if (currPageName != null) {
            TCAgent.onPageEnd(activity, currPageName);
        }
        currPageName = pageName;
        TCAgent.onPageStart(activity, currPageName);
    }

    @SuppressWarnings("unused")
    private void onPageBegin(final JSONArray args) throws JSONException {
        String pageName = args.getString(0);
        currPageName = pageName;
        TCAgent.onPageStart(talkingDataHTML.activity, currPageName);
    }
    
    @SuppressWarnings("unused")
    private void onPageEnd(final JSONArray args) throws JSONException {
        String pageName = args.getString(0);
        TCAgent.onPageEnd(talkingDataHTML.activity, pageName);
        currPageName = null;
    }
    
    @SuppressWarnings("unused")
    private void setLocation(final JSONArray args) {
        
    }
    
    private TDProfile.ProfileType intToProfileType(final int type) {
        switch (type) {
            case 0:
                return TDProfile.ProfileType.ANONYMOUS;
            case 1:
                return TDProfile.ProfileType.REGISTERED;
            case 2:
                return TDProfile.ProfileType.SINA_WEIBO;
            case 3:
                return TDProfile.ProfileType.QQ;
            case 4:
                return TDProfile.ProfileType.QQ_WEIBO;
            case 5:
                return TDProfile.ProfileType.ND91;
            case 6:
                return TDProfile.ProfileType.WEIXIN;
            case 11:
                return TDProfile.ProfileType.TYPE1;
            case 12:
                return TDProfile.ProfileType.TYPE2;
            case 13:
                return TDProfile.ProfileType.TYPE3;
            case 14:
                return TDProfile.ProfileType.TYPE4;
            case 15:
                return TDProfile.ProfileType.TYPE5;
            case 16:
                return TDProfile.ProfileType.TYPE6;
            case 17:
                return TDProfile.ProfileType.TYPE7;
            case 18:
                return TDProfile.ProfileType.TYPE8;
            case 19:
                return TDProfile.ProfileType.TYPE9;
            case 20:
                return TDProfile.ProfileType.TYPE10;
            default:
                return TDProfile.ProfileType.ANONYMOUS;
        }
    }
    
    private Map<String, Object> toMap(String jsonStr) {
        Map<String, Object> result = new HashMap<String, Object>();
        try {
            JSONObject jsonObj = new JSONObject(jsonStr);
            Iterator<String> keys = jsonObj.keys();
            String key = null;
            Object value = null;
            while (keys.hasNext()) {
                key = keys.next();
                value = jsonObj.get(key);
                result.put(key, value);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return result;
    }
    
    private Order orderFromJSONObject(JSONObject orderJson) {
        try {
            Order order = Order.createOrder(orderJson.getString("orderId"), orderJson.getInt("total"), orderJson.getString("currencyType"));
            JSONArray items = orderJson.getJSONArray("items");
            for (int i=0; i<items.length(); i++) {
                JSONObject item = items.getJSONObject(i);
                order.addItem(item.getString("itemId"), item.getString("category"), item.getString("name"), item.getInt("unitPrice"), item.getInt("amount"));
            }
            return order;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }
    
    private ShoppingCart shoppingCartFromJSONObject(JSONObject shoppingCartJson) {
        try {
            ShoppingCart shoppingCart = ShoppingCart.createShoppingCart();
            JSONArray items = shoppingCartJson.getJSONArray("items");
            for (int i=0; i<items.length(); i++) {
                JSONObject item = items.getJSONObject(i);
                shoppingCart.addItem(item.getString("itemId"), item.getString("category"), item.getString("name"), item.getInt("unitPrice"), item.getInt("amount"));
            }
            return shoppingCart;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }
}
